README.md
github-timetracking
===================

Chrome extension to add time tracking, reporting and billing to GitHub issues, with optional Pomodoro time tracking functionality. Also works as a standalone time tracking utility without GitHub.

It runs localy in your browser, and doesn't depend on any external service.

### Instalation

The production version is available from the Chrome Web Store at https://chrome.google.com/webstore/detail/github-timetracking/hbbbjnmncnckcoohigacfaifaiplacmo

To install the development version:

- Clone this repo
- Go to the extension page in Chrome
- Tick `Developer mode`
- Click `Load unpacked extension`
- Point to the directory where you cloned this repo


### Time Tracking

To start the timer click on the extension icon and press the play button. You can pause to stop the timer. To record the time spent in a Github issue, navigate to the issue in question and press `Record Time`. A new comment will be added to the issue, that will be parsed by the reports (see below).

#### Pomodoro functionality
The [Pomodoro Technique] (http://pomodorotechnique.com/) is a time management method to improve productivity and prevent burnout. Its principle is very simple: you work 25 mins and then take a rest for 5 mins. That's one pomodoro. After every 4 pomodori, take a long break of 30 mins instead of 5 mins.

To use it here, just tick `pomodoro intervals` before starting the timer. After 25 mins you'll see an alert and the counter will turn pink, counting the 5 mins short break. You have the ability to postone this break, or completely skip it. After the short break is finished another alert will sound and the red timer will resume. At the end of the 4th pomodori you'll see an alert and a green timer will start tracking the long break. Finally, after the long break is finished, the red timer will resume.

Note that the red timer will track only the billable time which includes the 25 mins of work time plus the 5 mins short breaks, not the long breaks. So after the long break finishes, the red timer will continue counting where it left off before the long break started.

Also note that you can disable the pomodoro intervals at any moment and return to regular time tracking by unticking the `pomodoro intervals` option.

### Reports

To generate a report, navigate first to the desired project in GitHub. Then click on `Project Report`. In the report pop-up select a milestone. You'll see the aggregated times per issue, and a grand total for the milestone. Click on `Send Bill` to create an email draft listing the issues and calculate the total amount due.

### Options
You can modify all the pomodoro durations by going into Chrome's extension page, and clicking on `options` in the GitHub TimeTracking section.


CHANGELOG.txt
version 2.0.0:
- Improved notifications.
- Added ability to postpone or completely skip breaks.
- Fixed: couldn't record time in github projects having dashes in their names.

version 1.1.0:
- Added options to be able to change default pomodoro times.


popup.js
$(function() {
  $('a.link').click(function() {
    chrome.tabs.create({url: $(this).attr('href')});
    return false;
  });
});

var background = chrome.runtime.getBackgroundPage(function(background) {
  chrome.tabs.query({'active': true, 'windowId': chrome.windows.WINDOW_ID_CURRENT}, function(tabs) {
      $('#no-project').show();
      $('#no-issue').show();
      $('#reports-show').attr('disabled', 'true');
      $('#record').attr('disabled', 'true');

      if (background.pomodoroEnabled) {
        $('input[name=pomodoro-intervals]').prop('checked', true);
        $('#pomodoros span').text(background.pomodori || "0");
        $('#pomodoros').show();
      } else {
        $('#pomodoros').hide();
      }

      var githubRe = /https:\/\/github.com\/([\w-]+)\/([\w-]+)(\/issues\/(\d+))?/;
      var currentUrl = tabs[0].url;
      var matches = currentUrl.match(githubRe);
      if (matches) {
        var account = matches[1];
        var project = matches[2];
        var issue = matches[4];
        
        if (typeof project != 'undefined') {
          chrome.tabs.executeScript({
            file: 'js/jquery-2.1.1.min.js'
          });
          chrome.tabs.executeScript({
            file: 'injected.js'
          });
          chrome.tabs.executeScript({
            file: 'js/bootstrap.js'
          });

          $('#no-project').hide();
          $('#project-name').html(project);
          $('#reports-show').removeAttr('disabled');

          if (issue) {
            $('#no-issue').hide();
            $('#issue').show();
            $('#record').removeAttr('disabled');
          } else {
            $('#issue').hide();
          }

          $('#reports-show').click(function() {
            $.get(
              'https://github.com/' + account + '/' + project + '/milestones',
              function(html) {
                html = $(html);
                $('select[name=milestone] option:not(:first)').remove();
                $('.milestone-title-link a', html).each(function() {
                  var href = $(this).attr('href');
                  $('select[name=milestone]')
                    .append($('<option value="' + href + '">' + $(this).text() + '</option>'));
                });

                chrome.tabs.sendMessage(tabs[0].id, {
                  action: 'SHOW_REPORTS',
                  project: project,
                  content: $('#reports-wrapper').html()
                });
              }
            );
          });
        }
      }

      if (background.status == background.STATUS_RUNNING) {
        $('#start').hide();
        $('#pause').show();
        $('#stop').show();
        $('#record').show();
      } else if (background.status == background.STATUS_PAUSED) {
        $('#start').show();
        $('#pause').hide();
        $('#stop').show();
        $('#record').show();
      }

      chrome.storage.sync.get({
        pomodoroDuration: 25,
        shortBreakDuration: 5,
        longBreaks: true,
        longBreaksEvery: 4,
        longBreakDuration: 30
      }, function(items) {
        background.optionsPomodoroDuration = items.pomodoroDuration;
        background.optionsShortBreakDuration = items.shortBreakDuration;
        background.optionsLongBreaks = items.longBreaks;
        background.optionsLongBreaksEvery = items.longBreaksEvery;
        background.optionsLongBreakDuration = items.longBreakDuration;
      });

      $('input[name=pomodoro-intervals]').change(function() {
        background.pomodoroEnabled = this.checked;
        if (this.checked) {
          $('#pomodoros').show();
        } else {
          $('#pomodoros').hide();
        }
      });

      $('#start').click(function() {
        background.startTimer();
        $('#start').hide();
        $('#pause').show();
        $('#stop').show();
        $('#record').show();
      });

      $('#pause').click(function() {
        background.pauseTimer();
        $('#start').show();
        $('#pause').hide();
      });

      $('#stop').click(function() {
        background.stopTimer();
        $('#start').show();
        $('#pause').hide();
        $('#stop').hide();
        $('#record').hide();
        $('#pomodoros span').text("0");
      });

      $('#record').click(function() {
        background.recordTime();
        $('#start').show();
        $('#pause').hide();
        $('#stop').hide();
        $('#record').hide();
      });
     }
  );

  function log(msg, arg) {
    if (typeof arg == 'undefined') {
      background.console.log(msg);
    } else {
      background.console.log(msg, arg);
    }
  }
});



injected.js
chrome.runtime.onMessage.addListener(function(req) {
  switch (req.action) {
    case 'COMMIT_TIME':
      document.querySelector('textarea[name=comment\\[body\\]]').value = 'Time spent: ' + req.time;
      document.querySelector('#partial-new-comment-form-actions button').click()
      break;
    case 'SHOW_REPORTS':
      $('.report').remove();
      $('body').append(req.content);
      $('#reports').modal('show');
      var milestoneTotalHours = milestoneTotalMinutes = 0;
      $('select[name=milestone]').change(function() {
        var milestoneHref = $(this).val();
        var parts = milestoneHref.split('/');
        var milestoneLabel = parts[parts.length - 1];
        $.get(
          'https://github.com/' + parts[1] + '/' + parts[2] + '/issues?q=milestone%3A' + encodeURIComponent(milestoneLabel),
          function(html) {
            html = $(html);
            $('.time-items').html('');
            $('.milestoneTotal').text('');
            milestoneTotalHours = milestoneTotalMinutes = 0;
            $('a.issue-title-link', html).each(function() {
              var issueTitle = $(this).text();
              var issueHref = $(this).attr('href');
              $.get(
                'https://github.com' + issueHref,
                function(html) {
                  var regex = /<p>Time spent: \d+:\d+<\/p>/g
                  var matches = html.match(regex);
                  if (matches) {
                    var totalHours = totalMinutes = 0;
                    $.each(matches, function(index, value) {
                      regex = /Time spent: (\d+):(\d+)/
                      matches2 = value.match(regex);
                      totalHours += parseInt(matches2[1]);
                      totalMinutes += parseInt(matches2[2]);
                    });
                    totalHours += Math.floor(totalMinutes / 60);
                    totalMinutes = totalMinutes % 60;
                    milestoneTotalHours += totalHours;
                    milestoneTotalMinutes = parseInt(milestoneTotalMinutes) + totalMinutes;
                    if (totalMinutes < 10) {
                      totalMinutes = "0" + totalMinutes;
                    }
                    $('#reports .time-items').append('<tr><td>' + issueTitle + '</td><td style="width:100px" class="text-right">' + totalHours + ':' + totalMinutes + '</td></tr>');
                    milestoneTotalHours += Math.floor(milestoneTotalMinutes / 60);
                    milestoneTotalMinutes = milestoneTotalMinutes % 60;
                    if (milestoneTotalMinutes < 10) {
                      milestoneTotalMinutes = "0" + milestoneTotalMinutes;
                    }
                    $('.milestoneTotal').text(milestoneTotalHours + ':' + milestoneTotalMinutes);
                  }
                }
              );
            });
          }
        );
      });

      $('#sendBill').click(function() {
        var rate = window.prompt('Please enter hourly rate for this project:');
        var recipient = window.prompt('Please enter email recipient:');
        var subject = "Invoice for project " + req.project;
        var amount = milestoneTotalHours * rate + milestoneTotalMinutes / 60 * rate;
        $('#reports .milestoneAmount').text("$" + parseFloat(amount).toFixed(2));

        var body = '';
        $('#reports tbody tr').each(function() {
          body += 'Task: "' + $('td:first', this).text().trim() + '"\r\n'
               + 'Time spent: ' + $('td:nth-child(2)', this).text().trim() + "\r\n\r\n";
        });

        body += '\r\nTotal Time: ' + $('#reports .milestoneTotal').text().trim();
        body += '\r\n\r\nAmount Due: ' + $('#reports .milestoneAmount').text().trim();

        body = "Reference milestone: " + $('select[name=milestone] option:selected').text()
          + "\r\n\r\n"
          + body;

        window.location.href = 'mailto:' + recipient + '?subject=' + encodeURIComponent(subject) + '&body=' + encodeURIComponent(body);
      });

      break;
  }
});


background.js
// CONSTANTS
var POMODORO_STATUS_WORK = 0;
var POMODORO_STATUS_REST_SHORT = 1;
var POMODORO_STATUS_REST_LONG = 2;
var STATUS_STOPPED = 0;
var STATUS_RUNNING = 1;
var STATUS_PAUSED = 2;

// EXTENSION OPTIONS
var optionsPomodoroDuration = 25;
var optionsShortBreakDuration = 5;
var optionsLongBreaks = true;
var optionsLongBreaksEvery = 4;
var optionsLongBreakDuration = 30;

// TIMESTAMPS
var initialTimestamp;
var currentPomodoroInitialTimestamp;
var originalBreakTimestamp;
var restTimestamp;
var nextBreakTimestamp;
var nextWorkTimestamp;

var intervalId;
var pomodoroEnabled = false;
var waitingAnswer = false;
var pomodoroStatus = POMODORO_STATUS_WORK;
var pomodori = 0;
var pausedAt = null;
var status = STATUS_STOPPED;

function tick() {
  var elapsedTime = getElapsedTime();
  var now = new Date();
  
  // currentPomodoroInitialTimestamp tracks pomodoros without postponed breaks, so it has
  // to be updated rigorously and independently of the rest events.
  if (now >= originalBreakTimestamp) {
    currentPomodoroInitialTimestamp = now;
    originalBreakTimestamp = datePlusMillis(currentPomodoroInitialTimestamp, (parseInt(optionsPomodoroDuration) + parseInt(optionsShortBreakDuration)) * 60000);
  }
  
  if (pomodoroEnabled && !waitingAnswer) {
    switch (pomodoroStatus) {
      case POMODORO_STATUS_WORK:
        showTime(elapsedTime[0], elapsedTime[1]);
        if (now >= nextBreakTimestamp) {
          new Audio('sounds/short-break.mp3').play();
          notification('break_1', isLongBreak()? "Time for a *LONG* break! (" + optionsLongBreakDuration + " mins)" : "Time for a *SHORT* break! (" + optionsShortBreakDuration + " mins)",
            [{title: 'Take a break'}, {title: 'Skip / Postpone'}]);
        }
        break;
      case POMODORO_STATUS_REST_SHORT:
      case POMODORO_STATUS_REST_LONG:
        showTime(elapsedTime[3], elapsedTime[4]);
        if (now >= nextWorkTimestamp) {
          setColor('#CC0000');
          if (pomodoroStatus == POMODORO_STATUS_REST_LONG) {
            // discount long breaks from billable time
            initialTimestamp = datePlusMillis(initialTimestamp, (optionsLongBreakDuration - optionsShortBreakDuration) * 60000)
          }
          pomodoroStatus = POMODORO_STATUS_WORK;
          nextBreakTimestamp = originalBreakTimestamp;
          newPomodoro();
          new Audio('sounds/work.mp3').play();
          break;
        }
    }
  } else {
    showTime(elapsedTime[0], elapsedTime[1]);
  }
}

function startTimer() {
  if (status == STATUS_PAUSED) {
    var pauseDuration = new Date() - pausedAt;
    initialTimestamp = datePlusMillis(initialTimestamp, pauseDuration);
    currentPomodoroInitialTimestamp = datePlusMillis(currentPomodoroInitialTimestamp, pauseDuration);
    originalBreakTimestamp = datePlusMillis(originalBreakTimestamp, pauseDuration);
    restTimestamp = datePlusMillis(restTimestamp, pauseDuration);
    nextBreakTimestamp = datePlusMillis(nextBreakTimestamp, pauseDuration);
    nextWorkTimestamp = datePlusMillis(nextWorkTimestamp, pauseDuration);
  } else {
    initialTimestamp = new Date();
    currentPomodoroInitialTimestamp = new Date();
    originalBreakTimestamp = datePlusMillis(currentPomodoroInitialTimestamp, optionsPomodoroDuration * 60000);
    restTimestamp = new Date();
    nextBreakTimestamp = nowPlusMillis(optionsPomodoroDuration * 60000);
  }
  setColor('#CC0000');
  tick();
  intervalId = setInterval(tick, 1000);
  status = STATUS_RUNNING;
}

function pauseTimer() {
  clearInterval(intervalId);
  pausedAt = new Date();
  setColor('#0066FF');
  status = STATUS_PAUSED;
}

function stopTimer() {
  clearInterval(intervalId);
  status = STATUS_STOPPED;
  pomodori = 0;
  chrome.browserAction.setBadgeText({
    text: ''
  });
}

function recordTime() {
  if (status > STATUS_STOPPED) {
    stopTimer();
  }
  var elapsedTime = getElapsedTime();
  chrome.tabs.query({'active': true, 'windowId': chrome.windows.WINDOW_ID_CURRENT}, function(tabs){
    chrome.tabs.sendMessage(tabs[0].id, {
      action: 'COMMIT_TIME',
      time: elapsedTime[0] + ':' + elapsedTime[1]
    });
    initialTimestamp = null;
    restTimestamp = null;
    nextBreakTimestamp = null;
  });
}

chrome.notifications.onButtonClicked.addListener(function(notificationId, buttonIx){
  chrome.notifications.clear(notificationId, function() {});
  waitingAnswer = false;
  switch (notificationId) {
    case 'break_1':
      if (buttonIx == 0) {
        // ***** Take a break
        restTimestamp = new Date();
        if (isLongBreak()) {
          setColor('#006600');
          pomodoroStatus = POMODORO_STATUS_REST_LONG;
          nextWorkTimestamp = nowPlusMillis(optionsLongBreakDuration * 60000);
        } else {
          setColor('#ff1493');
          pomodoroStatus = POMODORO_STATUS_REST_SHORT;
          nextWorkTimestamp = nowPlusMillis(optionsShortBreakDuration * 60000);
        }
      } else {
        var buttons = [{title: 'Skip'}];
        if (nowPlusMillis(2 * 60000) < originalBreakTimestamp) {
          buttons.push({title: 'Postpone'});
        }
        notification('break_2', 'Do you want to skip or postpone your break?', buttons);
      }
      break;
    case 'break_2':
      if (buttonIx == 0) {
        // ***** Skip break
        nextBreakTimestamp = originalBreakTimestamp;
        newPomodoro();
      } else {
        var buttons = [{title: '2 Minutes'}];
        if (nowPlusMillis(10 * 60000) < originalBreakTimestamp) {
          buttons.push({title: 'More'});
        }
        notification('break_3', 'For how long do you want to postpone your break?', buttons);
      }
      break;
    case 'break_3':
      if (buttonIx == 0) {
        // ***** Postpone break 2 minutes
        nextBreakTimestamp = nowPlusMillis(2 * 60000);
      } else {
        var buttons = [{title: '10 Minutes'}];
        if (nowPlusMillis(30 * 60000) < originalBreakTimestamp) {
          buttons.push({title: '30 Minutes'});
        }
        notification('break_4', 'For how long do you want to postpone your break?', buttons);
      }
      break;
    case 'break_4':
      if (buttonIx == 0) {
        // ***** Postpone break 10 minutes
        nextBreakTimestamp = nowPlusMillis(10 * 60000);
      } else {
        // ***** Postpone break 30 minutes
        nextBreakTimestamp = nowPlusMillis(30 * 60000);
      }
      break;
  }
});

function getElapsedTime() {
  var totalSecs = Math.floor((new Date() - initialTimestamp) / 1000);
  var restTotalSecs = Math.floor((new Date() - restTimestamp) / 1000);
  var totalMins = Math.floor(totalSecs / 60);
  var restTotalMins = Math.floor(restTotalSecs / 60);
  var totalHours = Math.floor(totalMins / 60);
  var restTotalHours = Math.floor(restTotalMins / 60);
  var mins = totalMins % 60;
  var restMins = restTotalMins % 60;
  var hours =  totalHours % 60;
  var restHours =  restTotalHours % 60;
  var seconds = totalSecs % 60;
  if (mins < 10) {
    mins = "0" + mins;
  }
  if (seconds < 10) {
    seconds = "0" + seconds;
  }
  if (restMins < 10) {
    restMins = "0" + restMins;
  }
return [hours, mins, seconds, restHours, restMins];
}

function showTime(hours, mins) {
  chrome.browserAction.setBadgeText({
    text: hours + ':' + mins
  });
}

function setColor(color) {
  chrome.browserAction.setBadgeBackgroundColor({
    color: color
  });
}

function notification(id, msg, buttons) {
  chrome.notifications.create(id, {
    type: 'basic',
    iconUrl: chrome.runtime.getURL('images/icon-48.png'),
    title: 'Time for a break',
    message: msg,
    priority: 2,
    buttons: buttons
  }, function() {});
  waitingAnswer = true;
}

function newPomodoro() {
  pomodori++;
}

function isLongBreak() {
  return (optionsLongBreaks && ((pomodori + 1 ) % optionsLongBreaksEvery == 0));
}

function datePlusMillis(date, millis) {
  return new Date(date.valueOf() + millis);
}

function nowPlusMillis(millis) {
  return new Date(new Date().valueOf() + millis);
}

// used for testing
function advanceTime(mins, secs) {
  secs = secs || 0;
  var diff = (mins * 60 + secs) * 1000;
  initialTimestamp = datePlusMillis(initialTimestamp, -diff);
  restTimestamp = datePlusMillis(restTimestamp, -diff);
  currentPomodoroInitialTimestamp = datePlusMillis(currentPomodoroInitialTimestamp, -diff);
  originalBreakTimestamp = datePlusMillis(originalBreakTimestamp, -diff);
  if (nextBreakTimestamp) {
    nextBreakTimestamp = datePlusMillis(nextBreakTimestamp, -diff);
  }
  if (nextWorkTimestamp) {
    nextWorkTimestamp = datePlusMillis(nextWorkTimestamp, -diff);
  }

  var t = getElapsedTime();
  console.log('clock: ', t[0], t[1], t[2]);
}


options.js
  var form = document.forms["settings"];

function save_options() {
  chrome.storage.sync.set({
    pomodoroDuration: form["pomodoroDuration"].value,
    shortBreakDuration: form["shortBreakDuration"].value,
    longBreaks: form["longBreaks"].checked,
    longBreaksEvery: form["longBreaksEvery"].value,
    longBreakDuration: form["longBreakDuration"].value
  }, function() {
    var status = document.getElementById('status');
    status.textContent = 'Options saved.';
    setTimeout(function() {
      status.textContent = '';
    }, 750);
  });
}

function restore_options() {
  chrome.storage.sync.get({
    pomodoroDuration: 25,
    shortBreakDuration: 5,
    longBreaks: true,
    longBreaksEvery: 4,
    longBreakDuration: 30
  }, function(items) {
    form["pomodoroDuration"].value = items.pomodoroDuration;
    form["shortBreakDuration"].value = items.shortBreakDuration;
    form["longBreaks"].checked = items.longBreaks;
    form["longBreaksEvery"].value = items.longBreaksEvery;
    form["longBreakDuration"].value = items.longBreakDuration;

    if (items.longBreaks) {
      document.getElementById('liLongBreaksEvery').style.display = 'list-item';
      document.getElementById('liLongBreaksDuration').style.display = 'list-item';
    }
  });
}
document.addEventListener('DOMContentLoaded', restore_options);
document.getElementById('save').addEventListener('click', save_options);

form["longBreaks"].addEventListener('click', function() {
  if (this.checked) {
    document.getElementById('liLongBreaksEvery').style.display = 'list-item';
    document.getElementById('liLongBreaksDuration').style.display = 'list-item';
  } else {
    document.getElementById('liLongBreaksEvery').style.display = 'none';
    document.getElementById('liLongBreaksDuration').style.display = 'none';
  }
});
